distributedlog-protocol/src/main/java/org/apache/distributedlog/EnvelopedRecordSetReader.java

```diff
@@ -115,4 +108,9 @@ public LogRecordWithDLSN nextRecord() throws IOException {
         return record;
     }
 
+    @Override
+    protected void finalize() throws Throwable {
```

Comment:
Just to be curious, why are you using finalize? 
I think that finalize is bad and legacy mechanism in Java and could lead in general to problems

Reply:
+1 on this, seems like its not something we should do lightly

how *are* you planning to solve it?